shared/src/pool_cache.rs

```diff
@@ -0,0 +1,524 @@
+use crate::{
+    current_block::{self, CurrentBlockStream},
+    maintenance::Maintaining,
+    pool_fetching::{Block, Pool, PoolFetching},
+};
+use anyhow::{Context, Result};
+use lru::LruCache;
+use model::TokenPair;
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    num::NonZeroU64,
+    sync::Mutex,
+};
+
+/// Caching pool fetcher
+///
+/// Caches all requests for a specific number of blocks and automatically updates the N most
+/// recently used pools automatically when a new block arrives.
+pub struct PoolCache {
+    mutexed: Mutex<Mutexed>,
+    number_of_blocks_to_cache: NonZeroU64,
+    inner: Box<dyn PoolFetching>,
+    block_stream: CurrentBlockStream,
+}
+
+// Design:
+// The design of this module is driven by the need to always return pools quickly so that end users
+// going through the api do not have to wait longer than necessary:
+// - The mutex is never held while waiting on an async operation (getting pools from the node).
```

Comment:
Do you mean it is never blocking or "locked"? Ahhh, you mean "held" like owned elsewhere.

Reply:
Technically mutexes can be locked and locks can be held so yeah the language here was a little inaccurate.